doc/src/guides/contributing/docstring.rst

```diff
@@ -156,7 +157,27 @@ is unnecessary, do not use it. Unnecessary sections and cluttered docstrings
 can make a function harder to understand. Aim for the minimal amount of
 information required to understand the function.
 
-1. Single-Sentence Summary
+1. Function Signature
+^^^^^^^^^^^^^^^^^^^^^
+
+The function signature is part of the Python source code but is almost always
+displayed along with the doc string. Function signatures should have
+informative argument names that follow Python variable name best practices,
+e.g. use ``coef_matrix`` instead of ``m``. It is also best to avoid the use of
+the generic ``*args`` and ``**kwargs`` in function signatures. Instead, make
+explicitly named arguments and keyword arguments as this is more informative for
+the user.
```

Comment:
I think what you really mean to say here is that functions should use explicit keyword arguments in the signature itself whenever they are known.

`*args` and `**kwargs` are the only way to do the things that they do. Any function that accepts an arbitrary number of arguments has to use `*args`, and there are a lot of those in SymPy (e.g., `Add`, `Mul`, `Piecewise`, etc.). Of course, if a function doesn't actually accept an arbitrary number of arguments, then `*args` is probably not correct, but I don't think this is a common enough thing that it needs to be mentioned here. 

`**kwargs` also also legitimately used whenever a function needs to pass arbitrary keyword arguments through. For example, `doit(**hints)` passes arbitrary hints through, and any hint can be defined by any Function subclass. 

I would also like to move toward using keyword-only arguments in more places, but the exact policy for that is harder to state (#21803).

Reply:
Happy to improve the wording.

Reply:
How about:

> Use explicit keyword arguments whenever they are known instead of `**kwargs`. For example, if a function takes a flag `simplify=True`, include `f(expr, simplify=True)` in the signature instead of using `f(expr, **kwargs)` with `kwargs.get('simplify', True)` in the body. Doing this will show the actual flag name and the default value in `help()` and the online documentation. If the default is something that can't be included in the function signature itself, use `None` instead and document what the default is in the docstring. `**kwargs` should only be used for functions that accept arbitrary keyword arguments, for example, `**assumptions` for functions that accept assumptions or `**hints` for functions like `doit()` that recursively pass hints through.

Again, I don't think we need to mention `*args` (unless you've seen people misusing it?). Ideally there would also be something there about making keyword arguments keyword-only, but we probably need to think a little more about what that policy should look like first.

Reply:
OK, I actually did just see this pattern in SymPy:

```py
def func(*args):
    x, y = args
```

So I guess people are misusing it :) `*args` is completely wrong there. If the number of args is a fixed number that is known ahead of time, the arguments should just be in the signature itself, like

```py
def func(x, y):
```

But in general, `*args` is actually needed in a lot of cases, so this needs to be worded in a way that only disallows `*args` for cases like the above. Or sometimes the number of arguments is a fixed number but it can be something like 2 args or 3 args. In those cases, it's sometimes simpler to just use `*args` and explain the signature in the docstring, rather than trying to emulate it without using `*args`. An example is `Range`, which accepts 1, 2, or 3 arguments, but the 1st argument is different for the 1-argument version, so trying to write it without `*args` would require something obtuse like `Range(start_or_stop, stop=None, step=None)`. Instead it just uses `Range(*args)` the the three versions are described at the top of the docstring.

Reply:
Range is an odd one that is designed to match `range` which also has a strange signature.

One thing I think that should be discouraged is mixing `*args` with accepting an iterable arg e.g.:
```python
In [14]: solve([x+y, x-y-1], [x, y])
Out[14]: {x(t): 1/2, y(t): -1/2}

In [15]: solve([x+y, x-y-1], x, y)
Out[15]: {x(t): 1/2, y(t): -1/2}
```
There isn't really a robust way to tell which form the user used. It is better just to be clear and ask for a single iterable argument. I think that `*args` should generally be discouraged unless there is some good reason for it like `Add(*args)`.

Reply:
> One thing I think that should be discouraged is mixing *args with accepting an iterable arg e.g.:

I agree with that (although which is better depends on the function). That's actually one benefit of type hints. When the types start getting complicated (like `Sequence[Expr] | Expr` instead of just `Expr`) it's a sign that there you might want to reconsider the API design. Although we do embrace polymorphic typing a lot in SymPy for the sake of usability (consider for instance, what the type signature of `subs` would be).